### ISSUE TITLE
FNX3-16111 ⁃ Pull-down bookmarks/history lists to trigger sync

### ISSUE THREAD

**Issue by sv-ohorvath** [mozilla-mobile/fenix#2165]
### Why/User Benefit/User Problem
It would be a simple, fast way to bring new synced items, without the need to navigate back & forth between account settings and library.

### What/Requirements
Pull-down to refresh and sync bookmarks & history.
Or, add a button in Bookmarks & history menus for that.

**Comment by ayushi1998** [comment]
Hey, I am an outreachy applicant. Can I work on this? How to go about the same? New to open source

**Comment by person808** [comment]
Gonna write down a couple of potential edge cases that I think I need UX feedback on.

Given this bookmark structure:

```
Mobile Bookmarks
|__A
   |__B
```

Edge cases:
1. If I'm in folder B, delete folder B on another device, sync that device, then trigger sync while in folder B, we end up viewing a folder that no longer exists, causing a crash.
2. If I'm in folder A, delete folder B on another device, sync that device, trigger a sync while in folder A, and navigate to folder B, then we are viewing a folder that no longer exists. I haven't tested yet, but this probably causes a crash.

Some potential solutions for 1:
1. Only allow the pull-to-refresh gesture when viewing folders that we know can't be deleted (ie Desktop Bookmarks, Booksmarks Toolbar, etc.) This is easy to implement but is a bit inconsistent.
2. If we sync and end up viewing an invalid folder, return to the top level bookmarks folder (Mobile Bookmarks). Harder to implement than 1, but I think still doable. Probably would need to show some kind of message indicating that the folder was removed?
3. Always go back to the top level bookmarks folder after syncing. Similar difficulty to implement as 2. Consistent behavior, but what if a user just wants to see a new bookmark from another device in a folder? Seems like having to navigate back to the folder would be annoying.

Some potential solutions for 2:
1. Cancel sync when opening a folder.
2. Implement one of solutions 2 or 3 from above.

Currently I have solution 1 implement for both problems as those are the easiest to implement but I believe the other solutions wouldn't be too much trouble. The only problem with the other solutions is the changes needed break both the back and up buttons so they would immediately go back to the previous screen instead of navigating up the bookmark tree. It looks like the back button is easy to fix, but I'm not sure about the up button.

**Comment by person808** [comment]
Now that I think about it, canceling sync when opening a folder is actually hard to get right because syncing and updating the bookmarks displayed are separate operations so we need to make sure we can't trigger sync and navigate before updating the bookmarks view.

**Comment by topotropic** [comment]
@apbitner can you have a look? Thanks!

**Comment by apbitner** [comment]
@person808 I would recommend solution #2. If the user tries to sync a folder that has been deleted elsewhere, we should return them to the parent (not necessarily top level) folder, where the deleted child folder should no longer be visible. 

> If I'm in folder A, delete folder B on another device, sync that device, trigger a sync while in folder A, and navigate to folder B, then we are viewing a folder that no longer exists.

For this edge case, after a sync has been triggered I would expect folder B would no longer be visible, so I'm not sure how the user could view it?

> If we try to edit a folder or bookmark, or we try to create a new folder while syncing, we might be making changes to a part of the bookmarks tree that will be invalid after the sync is finished.

I would assume the window in which a user could make changes while sync is occurring is so small that I would be okay if the user lost their changes, since they will be minor anyways.

**Comment by person808** [comment]
> For this edge case, after a sync has been triggered I would expect folder B would no longer be visible, so I'm not sure how the user could view it?

If the user taps on the folder before the sync completes then they can view the folder. Once the sync finishes, then the folder would be invalid. So the complete order of operations would be:

```
sync started -> user taps on folder B -> sync completes -> bookmarks are refreshed from device -> user is in invalid folder
```

It should be possible to check whether a folder is valid once the sync completes and return to the nearest valid parent folder if it isn't so I'll try that for now.

**Comment by person808** [comment]
Both gestures are merged! Just need qa to verify

**Comment by sflorean** [comment]
Verified as fixed on Nightly 6/30 and Beta 78.0.1-beta.2 with Samsung Galaxy Note 10 (Android 10) and LG G7 FIT (Android 8).
